### PR TITLE
[workflow]align the behavior of workflow's max_retires with remote function's max_retries

### PR DESCRIPTION
**Pull request by lchu6**
## Why are these changes needed?

To address the issue https://github.com/ray-project/ray/issues/22824

Basically the current behavior of `max_retries` in workflow is different from the one in remote functions in the following ways:
1. workflow's max_retries is not the number of retries, but the number of total tries. 
2. workflow's max_retries does not allow "-1" (infinite retries) while remote function's max_retries does.

This PR altered the behavior of `max_retries` in workflow to be consistent with the `max_retries` in remote functions:
1. make max_retries to be truly max retries (i.e. total tries = original try + max retries)
 - [x] implementation
 - [x] update logging
 - [x] update tests
2. make max_retries accept infinite tries (i.e. `max_retries=-1`)

## Related issue number

https://github.com/ray-project/ray/issues/22824

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/workflow/step_executor.py [hunk 1]
```diff
@@ -320,26 +320,50 @@ def _wrap_run(
...
-            f"\t[{i + 1}/{runtime_options.max_retries}]"
-        )
-        try:
-            result = func(*args, **kwargs)
-            exception = None
-            break
-        except BaseException as e:
-            if i + 1 == runtime_options.max_retries:
-                retry_msg = "Maximum retry reached, stop retry."
+    if runtime_options.max_retries != -1:
```
Comment (iycheng) [inline python/ray/workflow/step_executor.py:323]: let's do not duplicate the code here.

File: python/ray/workflow/tests/test_basic_workflows.py [hunk 2]
```diff
@@ -197,19 +197,16 @@ def unstable_step():
         return v
 
     with pytest.raises(Exception):
-        unstable_step.options(max_retries=-1).step().run()
```
Comment (iycheng) [inline python/ray/workflow/tests/test_basic_workflows.py:200]: This is testing invalid options are working. So let's use max_retries=-2

File: python/ray/workflow/step_executor.py [hunk 3]
```diff
@@ -318,28 +318,39 @@ def _wrap_run(
...
     # For ray failure, we should use max_retries.
-    for i in range(runtime_options.max_retries):
-        logger.info(
-            f"{get_step_status_info(WorkflowStatus.RUNNING)}"
-            f"\t[{i + 1}/{runtime_options.max_retries}]"
-        )
+    i = 0
+    while not done:
+        if i == 0:
+            logger.info(f"{get_step_status_info(WorkflowStatus.RUNNING)}")
```
Comment (iycheng) [inline python/ray/workflow/step_executor.py:327]:
```
            logger.info(f"{get_step_status_info(WorkflowStatus.RUNNING)}\tretries: [{i}]/unlimited")
```

**Review by fishbone — APPROVED** [review]
LG!

**Pull request merged**